### DISCUSSION TITLE
Facets: Streamlined Filtering and Advanced Search Integration for Directus Collections

### DISCUSSION THREAD

**Discussion by JayShoe** [directus/directus#23605] [Feature Requests]
Facets are a proposed schema-level feature for Directus, designed to centralize and simplify reusable filters, empowering administrators to create robust faceted browsing interfaces.

### Key Features:
- **Drill-Down Filtering**: Configure filters for attributes such as size, color, or price to enable intuitive, multi-level filtering experiences.  
- **Custom Search Behaviors**: Define tailored search logic and filter conditions to meet specific use cases.  
- **Advanced Search Integration**: Seamlessly connect with external search technologies like Elasticsearch, Algolia, or Solr for scalable and high-performance search capabilities.  

Facets allow search logic to be defined directly at the collection level in the Directus schema. These configurations are:  
- **Accessible via the Schema API**: Retrieve facet definitions for integration into applications.  
- **Applied via the Items API**: (GET, SEARCH, PATCH) Enable real-time data filtering.  
- **Summarized in Responses**: Display applied and available facets alongside query results.  
- **Extendable via the Extension API**: Define custom search parameters, synchronize remote search indexes, and merge external search results with application data.  

Facets integrate seamlessly with Directus Data Studio and custom front-end applications. They enable dynamic filter panels with interfaces such as checkboxes, radio buttons, sliders, and text inputs. Filters refine results in real time, while summaries provide clear insights into applied filters and remaining options.

This feature promises a consistent, scalable, and dynamic filtering experience across Directus systems and external platforms, enhancing usability and flexibility for administrators and end-users alike.

---
## **Basic Example**  

Here is a simple example using facets defined by the standard Directus filters. 

### **Example Schema**

This schema example demonstrates how facets can include predefined options for users to select, providing a structured and user-friendly filtering experience:

```json
{
  "collection": "products",
  "facets": {
    "category": {
      "label": "Category",
      "interface": "checkbox",
      "field": "category",
      "options": {
        "multiSelect": true,
        "choices": [
          { "value": "Clothing", "label": "Clothing" },
          { "value": "Electronics", "label": "Electronics" },
          { "value": "Accessories", "label": "Accessories" }
        ]
      }
    },
    "color": {
      "label": "Color",
      "interface": "checkbox",
      "field": "color",
      "options": {
        "multiSelect": true,
        "choices": [
          { "value": "Red", "label": "Red" },
          { "value": "Green", "label": "Green" },
          { "value": "Yellow", "label": "Yellow" },
          { "value": "Blue", "label": "Blue" }
        ]
      }
    },
    "price": {
      "label": "Price Range",
      "interface": "slider",
      "field": "price",
      "options": {
        "ranges": [
          { "from": 0, "to": 50, "label": "$0 - $50" },
          { "from": 51, "to": 100, "label": "$51 - $100" },
          { "from": 101, "to": 200, "label": "$101 - $200" }
        ]
      }
    }
  }
}
```

### **Explanation:**

1. **Category Facet**:
   - This facet uses a checkbox interface and provides predefined options, such as "Clothing," "Electronics," and "Accessories."
   - Users can select multiple options simultaneously due to the `multiSelect` setting.

2. **Color Facet**:
   - The color facet also uses a checkbox interface and offers predefined choices, including "Red," "Green," "Yellow," and "Blue."
   - Similar to the category facet, it supports multiple selections.

3. **Price Range Facet**:
   - The price facet uses a slider interface to filter by predefined price ranges.
   - Each range is labeled for clarity, such as "$0 - $50" or "$51 - $100."

---

### **Example Request**  
A simple drill-down facet filter request:  

```http
SEARCH  /items/products
{
  "facets": {
    "category": "Clothing",
    "color": "Red",
    "price": {
      "min": 20,
      "max": 50
    }
  }
}
```

- **`category`**: Filters products in the "Clothing" category.  
- **`color`**: Narrows results to "Red" items.  
- **`price`**: Limits results to products priced between $20 and $50.  

---

### **Example Response**  

```json
{
  "data": [
    {
      "id": 1,
      "title": "Red T-Shirt",
      "category": "Clothing",
      "color": "Red",
      "price": 25.00
    },
    {
      "id": 2,
      "title": "Red Hoodie",
      "category": "Clothing",
      "color": "Red",
      "price": 30.00
    }
  ],
  "facets": {
    "applied": {
      "category": "Clothing",
      "color": "Red",
      "price": {
        "min": 20,
        "max": 50
      }
    },
    "available": {
      "category": [
        { "value": "Clothing", "count": 10 },
        { "value": "Electronics", "count": 5 }
      ],
      "color": [
        { "value": "Red", "count": 15 },
        { "value": "Blue", "count": 8 },
        { "value": "Green", "count": 6 }
      ],
      "price_ranges": [
        { "from": 0, "to": 50, "count": 20 },
        { "from": 51, "to": 100, "count": 10 }
      ]
    }
  },
  "meta": {
    "total_count": 2,
    "page": 1,
    "page_size": 10
  }
}
```

**Explanation:**
1. **`data`**: Filtered results matching the applied facets.  
2. **`facets.applied`**: Shows filters used in the request.  
3. **`facets.available`**: Dynamically updates options based on results:
   - Remaining categories, colors, and price ranges with counts.  
4. **`meta`**: Pagination details.  

---

### **How It Works**  
- The Schema API defines facets for front-end filtering panels.  
- Facets are linked to collection fields with optional settings (e.g., `multiSelect`, ranges).  
- The front end dynamically generates interfaces (e.g., checkboxes, sliders) and applies filters for consistent and user-friendly behavior.  
- The results are filtered by the applied facets, and summaries are included in the results.

This schema-driven approach ensures seamless integration across Directus Data Studio and custom applications.

---

## **Advanced Example**  

This advanced example demonstrates how a text box facet can be configured for the front-end to act as a search box. This advnaced functionality would be implemented in "phase 2" of the facets feature. This example would be implemented by the community via an extension. Directus simply creates the abstraction layer so that the extension can interact with the Directus API . 

The applied facet relies on an Elasticsearch extension to form and execute the query. The facet definition defines which fields are synchronized with the external search index to ensure efficient and accurate results. The extension takes the query, reformats and submits the query to the external search engine, merges the data with internal Directus data, and returns them for the front-end to display.

---

### **Example Schema**

Below is a schema configuration for a text box facet tied to an Elasticsearch index:

```json
{
  "collection": "products",
  "facets": {
    "search_box": {
      "label": "Search Box",
      "interface": "text",
      "field": "search_query",
      "sync": {
        "index": "elasticsearch",
        "endpoint": "https://example-elasticsearch.com/products",
        "fields": ["title", "description", "tags"]
      }
    }
  }
}
```

**Explanation:**
- **`sync.index`**: Specifies the external search engine (Elasticsearch in this case).
- **`sync.endpoint`**: Defines the Elasticsearch endpoint used for synchronization.
- **`sync.fields`**: Lists the collection fields synchronized with the external index.

In this schema:
- The `title`, `description`, and `tags` fields from the `products` collection are synchronized with the Elasticsearch index.
- The `search_box` facet is exposed to the front-end as a text box for user input.

---

### **Example Request**

The following request uses the text box facet to perform a search:

```http
POST /items/products/search
{
  "facets": {
    "search_box": "Wireless Headphones"
  }
}
```

**Explanation:**
- The `search_box` facet allows the user to input "Wireless Headphones."
- The Elasticsearch extension converts this input into a query for the external index.

---

### **Elasticsearch Query Translation**

The extension translates the `search_box` input into an Elasticsearch-compatible query:

```json
{
  "query": {
    "multi_match": {
      "query": "Wireless Headphones",
      "fields": ["title", "description", "tags"]
    }
  }
}
```

---

### **Example Response**

The merged results from Elasticsearch and Directus data are returned as follows:

```json
{
  "data": [
    {
      "id": 1,
      "title": "Wireless Noise-Canceling Headphones",
      "description": "High-quality sound with active noise cancellation.",
      "tags": ["audio", "headphones", "wireless"],
      "price": 150.00,
      "stock": 25
    },
    {
      "id": 2,
      "title": "Wireless Over-Ear Headphones",
      "description": "Comfortable fit and excellent sound quality.",
      "tags": ["headphones", "audio", "wireless"],
      "price": 100.00,
      "stock": 15
    }
  ],
  "facets": {
    "applied": {
      "search_box": "Wireless Headphones"
    },
    "available": {
      "search_box": ["Wireless Earbuds", "Wireless Headphones", "Bluetooth Speakers"]
    }
  },
  "meta": {
    "total_count": 2,
    "page": 1,
    "page_size": 10
  }
}
```

**Explanation:**
- **`data`**: Displays the merged results from Elasticsearch and Directus, showing only items relevant to the query.
- **`facets.applied`**: Reflects the input provided in the search box.
- **`facets.available`**: Suggests related terms or commonly searched items based on the external search engine's capabilities.
- **`meta`**: Includes pagination and summary details for the results.

---

This example demonstrates how facets facilitate seamless integration with external search engines, enabling dynamic and efficient filtering while preserving consistency with Directus's internal data schema. However, this advanced functionality would be part of phase 2 in the development of the Facets feature. Establishing this abstraction layer early ensures the feature is designed with scalability and future integration in mind during its initial implementation.

---

### **Motivation**  

Facets tackle key challenges in managing filters and queries in Directus by providing a centralized, reusable, and intuitive way to define and apply filters across collections. This approach simplifies configuration, enhances manageability, and enables dynamic filtering for diverse use cases.

#### **Simplifying Relational Data Queries**  
Facets address the complexity of many-to-many (M2M) and many-to-one (M2O) relationships in Directus. These queries are often difficult to construct and maintain, especially for non-technical users. By abstracting these relationships into schema-defined facets, the system reduces complexity and enables intuitive filtering for relational data.

#### **Integrating Advanced Search Engines**  
Facets create an abstraction layer for seamless integration with advanced search engines such as Elasticsearch, Algolia, or Solr. Each search endpoint can be configured within a facet definition, with conditions (e.g., roles, users, or contextual factors) determining its exposure. Additionally, facets manage synchronization and updates to external search indexes, allowing front-end applications to perform efficient, context-aware queries.

---

### **Use Cases**
1. **Faceted Browsing:**  
   Facilitate intuitive filtering for e-commerce and similar applications, enabling users to filter items by attributes like size, color, or price.

2. **Advanced Filtering Logic:**  
   Show or hide facets based on user permissions, or contextual conditions (current category, applied facets, etc).

3. **Dynamic Updates:**  
   Automatically update available options based on filtered results (e.g., showing only applicable colors within a selected product category).

4. **Search Engine Integration:**  
   Use facets as an abstraction layer to generate query formats required by tools like Elasticsearch or Algolia, simplifying the integration process.

### **Expected Outcome**  
Facets empower administrators to define flexible, reusable search logic in Directus, streamlining both internal filtering processes and external integrations. This approach reduces complexity, enhances usability for non-technical users, and unlocks advanced search capabilities across diverse platforms. By centralizing filters, facets enable scalable, intuitive, and efficient data navigation and searches for all user levels.

---

#### **1. Facet Configuration in Schema**  
Facets are defined in the Directus schema to ensure consistency and reusability across collections. Each facet includes:  
- **Target Field(s)**: Specifies the field(s) to which the facet applies.  
- **Interface Type**: Defines how the facet is displayed, such as a slider, checkbox, or text input.  
- **Additional Settings**: Supports configurations like predefined ranges, geospatial constraints, or validation rules.  
- **Conditional Visibility**: Allows facets to be shown or hidden based on other facets, user roles, or contextual factors.  
- **Extension-Specific Settings**: Includes parameters for external indexes and search endpoint locations for integration with advanced search engines.

#### **2. Facet Settings UI**  
Facets are configured in the **Collection Settings**, providing administrators with a user-friendly interface to:  
- Define facet properties (e.g., target fields, ranges, validation).  
- Manage display order through a sortable list.  
- Easily configure facets using an interface similar to the existing field configuration pane.  

#### **3. Facets Sidebar for Faceted Browsing in Directus Studio**  
Facets are presented in the **information sidebar** of the Directus Studio, enabling dynamic filtering of results based on user input. Examples include:  
- **Checkboxes**: For multi-select filtering by attributes like size or color.  
- **Sliders**: For range filtering, such as price or numerical values.  
- **Text Inputs**: For freeform searches tied to advanced search engines or specific configurations.  

The sidebar dynamically updates to reflect the applied filters and the available filtering options based on the current dataset.

#### **4. Advanced Integration with Search Engines**  
Facets act as an abstraction layer, allowing Directus to define reusable queries that can be translated into formats compatible with advanced search engines like Elasticsearch, Algolia, or Solr. Key benefits include:  
- **Multiple Search Definitions per Collection**: Administrators can define different search behaviors tailored to specific use cases within the same collection.  
- **Customizable Query Logic**: Extensions translate facet definitions into external search engine queries while merging results with Directus data.  
- **Flexible Use Cases**: Supports various scenarios, from basic filtering within Directus to highly specialized searches using external indexes.  

This abstraction ensures scalability, enabling administrators to configure complex search logic without requiring extensive technical expertise.

---

#### **Must Have:**
- Schema-level configuration for reusable and extensible facets.
- A Facet Settings UI in the Collection Settings to define, manage, and organize facets.
- API support to retrieve facet definitions and dynamically apply facet-based queries.
- Integration of faceted browsing functionality into the information sidebar for end-users.

#### **Should Have:**
- A `facets` extension type to enable integration with database specific features such as POSTGIS in Postgres, and advanced search engines like Elasticsearch or Algolia.
- Real-time hooks to ensure external search indexes are synchronized with Directus data changes.
- Support for geospatial facets and multi-field facet configurations.
- Dynamic updates to facet availability based on filtered results.
- Optional hierarchical facets for complex filtering (e.g., "Category > Subcategory").
- Configurable display templates to customize the appearance of facets (e.g., text, icons).
- Translations support to display interfaces in the specified language. 

#### **Could Have:**
- Built-in analytics to track facet usage and popular filters.
- Enhanced accessibility support, such as ARIA attributes and screen reader compatibility.
- A built-in "preview mode" to test facets and validate their functionality during setup.

#### **Won’t Have:**
- Database-specific implementations tied directly to SQL or NoSQL technologies.  
- Predefined integrations with third-party search engines; instead, this will be handled through extensions.

---

### **Implementation Strategy**

The initial implementation should focus on basic Directus-based filtering, enabling faceted searches using existing filter capabilities. Priority should be given to building the foundational scaffolding that supports schema definitions and UI components. This will ensure a seamless transition for extensions to leverage the framework for advanced search integrations, such as Elasticsearch or Algolia, in a planned second phase.

---

### **Alternatives**
1. **Custom User-Space Solutions**: Developers could build their own faceted search interfaces using the Directus Extensions API. While this offers full control over the design, it requires more effort and technical expertise.
2. **Improved Default Filtering UI**: Enhance the existing filtering system with autocomplete, better UX for multiple filters, and improved documentation for non-technical users. The challenge here is to be able to shoehorn various search technologies into the existing methods. 
3. **Impact of Not Doing This**: Without facets, less technical users may struggle with filtering large datasets. Developers will need to implement custom solutions, which are less efficient and harder to maintain compared to a core feature.

---

### **Adoption Strategy**
1. **Non-Breaking**: Facets are an optional, non-breaking feature that won’t disrupt existing setups.
2. **Opt-In**: Developers can configure predefined facets via the admin UI without mandatory changes.
3. **No Migration Needed**: Existing projects won’t require migration; facets can be added as needed.
4. **Minimal Impact**: Improves usability for data-heavy projects without affecting other features.

---

### **Drawbacks**  
1. **Implementation Complexity**: Requires updates to schema, UI, API, and extensions.
2. **Index Maintenance**: External search engines need synchronization with Directus data changes.
3. **Learning Curve**: Administrators need to understand how to configure and manage facets effectively.

---

### **Unresolved Questions**
1. Can we combine results from an external search engine, as well as standard filters? If not, we must add conditions to use one method or the other. 
~~2. There is a blocking issue that prevents multiple M2M, M2O and O2M relationships from working in general. https://github.com/directus/directus/issues/20614~~

---

Facets bring an easy, flexible way to manage filters in Directus, whether you’re just setting up basic browsing or connecting to advanced search engines like Elasticsearch. They simplify workflows, make filtering more intuitive for everyone, and open up powerful possibilities for scaling your projects. This feature has the potential to make Directus even more user-friendly and adaptable to a wide range of needs.

**Comment by ImaCrea** [comment]
💯 !! Would be a game changer. Our users are overloaded with too many options in their filters at the moment because of all the relationship and being able to serve them only what they need would be a massive improvement to them.

**Comment by rijkvanzanten** [comment]
Big fan. Great idea! 🙂 Thanks @JayShoe

**Reply by JayShoe** [reply]
Thank you @rijkvanzanten. I'm going to further define this feature below.

**Comment by JayShoe** [comment]
NOTE: I edited the main summary to incorporate additional information regarding the API integrations and detailed design.

**Reply by JayShoe** [reply]
NOTE: I edited the main summary to incorporate information regarding advanced search engine integration and how this would fit into the facets concept.

**Comment by lpkobamn** [comment]
It's going to be incredibly cool, thank you for describing it.

**Comment by salva-dev** [comment]
Great feature!!

**Comment by JayShoe** [comment]
Now that https://github.com/directus/directus/issues/20614 is resolved, this feature is more viable.
